addons/tls13-middlebox/bootstrap.js

```diff
@@ -0,0 +1,276 @@
+      });
+
+      xhr.addEventListener("timeout", e => {
+        reportResult("timeout", e.target);
+      });
+
+      xhr.send();
+    } catch (ex) {
+      resolve(Object.assign({result: {"event": "exception", "description": ex.toSource()}}, config));
+    }
+  });
+}
+
+// shuffle the array randomly
+function shuffleArray(original_array) {
+  let copy_array = original_array.slice();
+
+  let output_array = [];
+
+  while (copy_array.length > 0) {
+    let x = Math.floor(Math.random() * copy_array.length);
+    output_array.push(copy_array.splice(x, 1)[0]);
+  }
+
+  return output_array;
+}
+
+// make the request for each configuration
+async function runConfigurations() {
+  let result = [];
+
+  for (let config of shuffleArray(configurations)) {
+    // we wait until the result is ready for the current configuration
+    // and then move on to the next configuration
+    result.push(await makeRequest(config));
+  }
+
+  return result;
+}
+
+// check if either of VERSION_MAX_PREF or FALLBACK_LIMIT_PREF was set by the user
+function hasUserSetPreference() {
+  let readonly_prefs = new Preferences();
+
+  if (readonly_prefs.isSet(VERSION_MAX_PREF) || readonly_prefs.isSet(FALLBACK_LIMIT_PREF)) {
+    // reports the current values as well as whether they were set by the user
+    isNonBuiltInRootCertInstalled().then(non_builtin_result => {
+      TelemetryController.submitExternalPing("tls13-middlebox", {
+        maxVersion: {
+          value: readonly_prefs.get(VERSION_MAX_PREF),
+          isUserset: readonly_prefs.isSet(VERSION_MAX_PREF)
+        },
+        fallbackLimit: {
+          value: readonly_prefs.get(FALLBACK_LIMIT_PREF),
+          isUserset: readonly_prefs.isSet(FALLBACK_LIMIT_PREF)
+        },
+        isNonBuiltInRootCertInstalled: non_builtin_result
+      });
+
+      return true;
+    }).catch();
+
+    return true;
+  }
+
+  return false;
+}
+
+function startup() {}
+
+function shutdown() {}
+
+function install() {
+  // abort if either of VERSION_MAX_PREF or FALLBACK_LIMIT_PREF was set by the user
+  if (hasUserSetPreference()) {
+    return;
+  }
+
+  // record the default values before the experiment starts
+  let defaultMaxVersion = readwrite_prefs.get(VERSION_MAX_PREF);
+  let defaultFallbackLimit = readwrite_prefs.get(FALLBACK_LIMIT_PREF);
+
+  runConfigurations().then(tests_result => {
+    // restore the default values after the experiment is over
+    readwrite_prefs.set(VERSION_MAX_PREF, defaultMaxVersion);
+    readwrite_prefs.set(FALLBACK_LIMIT_PREF, defaultFallbackLimit);
+
+    // report the test results to telemetry
+    isNonBuiltInRootCertInstalled().then(non_builtin_result => {
+      TelemetryController.submitExternalPing("tls13-middlebox-v1", {
+        "defaultMaxVersion": defaultMaxVersion,
+        "defaultFallbackLimit": defaultFallbackLimit,
+        "isNonBuiltInRootCertInstalled": non_builtin_result,
+        "tests": tests_result
+      });
+
+      return true;
+    }).catch();
```

Comment:
As above I'd rather see an optional log message or something rather than a bare `catch`, for local testing purposes if nothing else...